Cpp code genrators for ROS2 nodes

ipa-nhg opened the pull request:
These changes creates for the RosModel model:

```
PackageSet { 
	CatkinPackage test_ros2{ Artifact  test_ros2 {
	  Node { name test_ros2 
		Publishers {
			Publisher { name string_pub message "std_msgs.String"},
			Publisher { name battery_pub message "sensor_msgs.BatteryState"}}
		Subscribers {
			Subscriber {name bool_sub message "std_msgs.Bool"},
			Subscriber {name image_sub message "sensor_msgs.Image"}}
		}
	}}}
```

the following ROS package:

https://github.com/ipa-nhg/test_ros2_code_extractor/tree/master/test_ros2



To test it I used the [`ros-model-extractors`](https://github.com/ipa320/ros-model-extractors) and its foxy docker container with the commands:

```
docker build --tag=haros_foxy -f foxy/Dockerfile .
docker run -it haros_foxy:latest /haros_runner.sh test_ros2 test_ros2 node . /root/ws https://github.com/ipa-nhg/test_ros2_code_extractor
```

The resulted model was:

```
## Init HAROS ##
[HAROS] Running setup operations...

## Call the HAROS plugin to extract the ros-models ##
###########
~~~~~~~~~~~
Print of the model: test_ros2.ros:
~~~~~~~~~~~
PackageSet {
  CatkinPackage test_ros2 { 
    Artifact test_ros2 {
      Node { name test_ros2
        Publishers {
          Publisher { name 'string_pub' message 'std_msgsString'},
          Publisher { name 'battery_pub' message 'sensor_msgsBatteryState'}}
        Subscribers {
          Subscriber { name 'bool_sub' message 'std_msgsBool'},
          Subscriber { name 'image_sub' message 'sensor_msgsImage'}}}}
}}
~~~~~~~~~~~
###########
```

ipa-nhg commented:
Comments from @ipa-hsd of potential improvements for the code template:

- [ ] I would avoid having a main() in the node cpp. maybe a separate main.cpp with all the nodes in that package are added to executor. (this would be just an example). https://github.com/ros2/examples/blob/master/rclcpp/executors/multithreaded_executor/multithreaded_executor.cpp#L167-L175
- [ ] Instead use nodes as components: https://github.com/ros2/demos/blob/master/composition/src/listener_component.cpp#L55
	this will give and option to load nodes as components (available only for rclcpp)
	https://docs.ros.org/en/foxy/Tutorials/Composition.html
- [ ] Quality attributes can be used for Qos? This one uses only history QoS (with default 10)
	https://github.com/ipa320/ros-model/blob/master/plugins/de.fraunhofer.ipa.roscode.generator/src/de/fraunhofer/ipa/roscode/generator/Ros2CodeGenerator.xtend#L74
	Example: https://github.com/ros2/demos/blob/master/image_tools/src/cam2image.cpp#L65-L81
	https://github.com/ros2/demos/blob/master/image_tools/src/cam2image.cpp#L65-L81
- [ ] Parameters and actions are missed

ipa-hsd commented on plugins/de.fraunhofer.ipa.ros.xtext.tests/resources/ros2generator/test_pkg/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+find_package(sensor_msgs REQUIRED)
+
+add_executable(test_node src/test_node.cpp)
+ament_target_dependencies(test_node std_srvs sensor_msgs )
```
rclcpp is missing I think

ipa-nhg commented:
I updated here also the ROS1 code generator to generate also a full package.

For testing I added the autogenerated example to: https://github.com/ipa-nhg/test_ros_code_extractor and I am using the cpp code that we are generating to test the extractors: https://github.com/ipa320/ros-model-extractors/commit/dbfc5a6506978e54f775e7fbca6b5c93e039f895#diff-3722d9ba8feb2d3feac8ce71a209a638d4b404e1c53f937188761181594023e2R23


All the tests are successful:  https://github.com/ipa320/ros-model-extractors/actions 👏🏽 👏🏽 

@ipa-hsd thanks for your help! should we merge this version and continue the improvements of the ROS2 code generation in https://github.com/ipa320/ros-model/pull/135 ?

ipa-hsd commented:
LGTM let's merge this one!
